### DISCUSSION TITLE
Can we use `Depends` as a decorator?

### DISCUSSION THREAD

**Discussion by winstxnhdw** [fastapi/fastapi#9569] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import AsyncGenerator

from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession

from server.database import Database


@Depends
async def get_database() -> AsyncGenerator[AsyncSession, None]:
    """
    Summary
    -------
    FastAPI dependency that returns a database session

    Returns
    -------
    session (AsyncSession) : database session
    """
    async with Database.AsyncSessionLocal() as session:
        yield session
```

### Description

I have been using `Depends` as a decorator instead of calling it in my parameters.

```python
@v1.get('/register', status_code=HTTP_201_CREATED)
async def register(request: User, session: AsyncSession = get_database):
    """
    Summary
    -------
    the route to handle user registration
    """
    try:
        await create_user(session, request.email, request.password)

    except IntegrityError as exception:
        await session.rollback()
        raise HTTPException(HTTP_409_CONFLICT, "User already exists!") from exception
```

I have not faced any issues with it yet as the resulting function is technically(?) the same. I have not seen anyone do this so I could very well be misusing `Depends` and have yet to hit some edge case. Any thoughts?

### Operating System

Linux

### Operating System Details

Arch Linux

### FastAPI Version

0.95.2

### Python Version

3.11.3

**Comment by Kludex** [comment]
Can you use `app.dependency_overrides`?

**Reply by winstxnhdw** [reply]
I am not sure what you mean.

**Reply by Kludex** [reply]
```python
from fastapi import FastAPI, Depends

app = FastAPI()

@Depends
def potato():
    return 1

@app.get("/")
def home(d = potato):
    return d

app.dependency_overrides = {potato: lambda: 2}
```

Does this work as expected?

**Reply by winstxnhdw** [reply]
Nope, it still returns `1`. Guess we can't use `Depends` as a decorator. I wonder why.

**Reply by DJRHails** [reply]
I think it's because the `dependency_overrides` (and a lot of other code) explicitly looks at the type annotations (with a class of `Depends`) to determine dependencies.
